File: 08.py
# https://leetcode.cn/problems/zero-matrix-lcci/
from typing import List


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        rows, cols = len(matrix), len(matrix[0])
        # 收集0元素坐标
        zeros = []
        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == 0:
                    zeros.append((i, j))

        zeros[::] = list(zip(*zeros))
        if zeros:
            for i in zeros[0]:
                for j in range(cols):
                    matrix[i][j] = 0

            for j in zeros[1]:
                for i in range(rows):
                    matrix[i][j] = 0


if __name__ == '__main__':
    test = Solution()
    matrix = [
        [0, 1, 2, 0],
        [3, 4, 5, 2],
        [1, 3, 1, 5]
    ]
    matrix = [[1]]
    test.setZeroes(matrix)
    print(matrix)

File: Interv01/08.cpp
//
// Created by Air on 2023/2/27.
//
// https://leetcode.cn/problems/zero-matrix-lcci/

#include "iostream"
#include "vector"
#include "unordered_set"
using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int rows(matrix.size()), cols(matrix[0].size());
        // 收集0元素坐标
        unordered_set<int> zero_rows, zero_cols;
        for (int i = 0; i < rows; ++i)
            for (int j = 0; j < cols; ++j)
                if (!matrix[i][j]) {
                    zero_rows.insert(i);
                    zero_cols.insert(j);
                }

        if (zero_rows.size() && zero_cols.size()) {
            for (int i : zero_rows)
                for (int j = 0; j < cols; ++j)
                    matrix[i][j] = 0;
            for (int i : zero_cols)
                for (int j = 0; j < rows; ++j)
                    matrix[j][i] = 0;
        }

    }
};

int main() {
    vector<vector<int>> matrix = {
            {0,1,2,0},
            {3,4,5,2},
            {1,3,1,5}
    };
    Solution test;
    test.setZeroes(matrix);
    int rows(matrix.size()), cols(matrix[0].size());
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}
